README.md
QMLeteroid
=====

Python Code Health: [![Python Code Health](https://landscape.io/github/YtvwlD/qmleteroid/master/landscape.png)](https://landscape.io/github/YtvwlD/qmleteroid/master)


a QML clone of https://github.com/chaosdorf/meteroid

## Setup ##

```
sudo apt install qmlscene pyotherside python3-requests
```

## Run ##

`qmlscene main.qml`


debian/changelog
qmleteroid (0.1-1) xenial; urgency=medium

  * Initial release.

 -- Niklas Sombert <niklas@ytvwld.de>  Wed, 15 Jun 2016 13:37:37 +0200


lib.py
# QMLeteroid - a QML clone of https://github.com/chaosdorf/meteroid
# Copyright (C) 2016 Niklas Sombert
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

from json import JSONDecoder
from hashlib import md5
import requests

json = JSONDecoder()

def get_users_data(url):
	req = requests.get("{}/users.json".format(url))
	content = json.decode(req.text)
	results = []
	for element in content:
		results.append(_parseUser(element))
	return results

def get_user_data(url, uid):
	req = requests.get("{}/users/{}.json".format(url, uid))
	content = json.decode(req.text)
	assert content["id"] == uid
	return _parseUser(content)

def save_user_data(url, uid, name, balance, email):
	p = {
		"user[name]": name,
		"user[email]": email,
		"user[balance]": balance
	}
	requests.patch("{}/users/{}.json".format(url, uid), params=p)

def add_user(url, name, balance, email):
	p = {
		"user[name]": name,
		"user[balance]": balance,
		"user[email]": email
	}
	req = requests.post("{}/users.json".format(url), params=p)
	content = json.decode(req.text)
	return _parseUser(content)

def _parseUser(data):
	return {
		"id": data["id"],
		"name": data["name"],
		"balance": data["balance"],
		"email": data["email"],
		"portrait": get_gravatar_url(data["email"])
	}

def get_gravatar_url(email):
	return "http://gravatar.com/avatar/{}".format(md5(email.strip().lower().encode()).hexdigest())

def get_drinks(url):
	req = requests.get("{}/drinks.json".format(url))
	content = json.decode(req.text)
	results = []
	for element in content:
		results.append({
			"id": int(element["id"]),
			"name": element["name"],
			"donation_recommendation": element["donation_recommendation"],
			"logo": ("{}/{}".format(url, element["logo_url"])) if element["logo_url"] else ""
		})
	return results

def buy_drink(url, uid, did):
	requests.get("{}/users/{}/buy?drink={}".format(url, uid, did))


lib2.py
# QMLeteroid - a QML clone of https://github.com/chaosdorf/meteroid
# Copyright (C) 2016 Niklas Sombert
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

fake_users_list = [
	{
		"id": 0,
		"name": "A",
		"balance": 10,
		"email": "mail@example.com",
		"portrait": "/usr/share/ubuntu-html5-ui-toolkit/0.1/ambiance/img/avatar_contacts_list@8.png",
	},
	{
		"id": 1,
		"name": "B",
		"balance": 5,
		"email": "mail@example.net",
		"portrait": "/usr/share/ubuntu-html5-ui-toolkit/0.1/ambiance/img/avatar_contacts_list@8.png",
	},
	{
		"id": 2,
		"name": "C",
		"balance": 2.50,
		"email": "mail@example.org",
		"portrait": "/usr/share/ubuntu-html5-ui-toolkit/0.1/ambiance/img/avatar_contacts_list@8.png",
	},
]

def get_users_data(url):
	return fake_users_list

def get_user_data(url, uid):
	return fake_users_list[uid]

def save_user_data(url, uid, name, balance, email):
	print("This is not real.")

def add_user(url, name, balance, email):
	print("This is not real.")

def get_drinks(url):
	return [{
		"id": 0,
		"name": "Mate",
		"bottle_size": 0.5,
		"caffeine": True,
		"donation_recommendation": 1.5,
		"logo": ""
	}]

def buy_drink(url, uid, amount):
	print("This is not real.")


BuyDrink.py
from json import JSONDecoder
from hashlib import md5
import requests

json = JSONDecoder()

def get_user_info(uid):
	req = requests.get("http://mete/users/{}.json".format(uid))
	content = json.decode(req.text)
	assert content["id"] == uid
	return {
		"name": content["name"],
		"id": content["id"],
		"balance": content["balance"]
	}

def get_drinks():
	req = requests.get("http://mete/drinks.json")
	content = json.decode(req.text)
	results = []
	for element in content:
		results.append({
			"id": int(element["id"]),
			"name": element["name"],
			"bottle_size": element["bottle_size"],
			"caffeine": element["caffeine"],
			"donation_recommended": element["donation_recommended"],
			"logo": ("http://mete/{}".format(element["logo_url"])) if element["logo_url"] else ""
		})
	return results

def buy_drink(uid, amount):
	req = requests.get("http://mete/users/{}/deposit?amount=-{}".format(uid, amount))

def get_user_info2(uid): #offline debugging
	return {
		"name": "A",
		"id": 0,
		"balance": 10
	}

def get_drinks2(): #offline debugging
	return [{
		"id": 0,
		"name": "Mate",
		"bottle_size": 0.5,
		"caffeine": True,
		"donation_recommended": 1.5,
		"logo": ""
	}]

def buy_drink2(uid, amount):
	pass


PickUsername.py
from json import JSONDecoder
from hashlib import md5
import requests

def get_data():
	json = JSONDecoder()
	req = requests.get("http://mete/users.json")
	content = json.decode(req.text)
	results = []
	for element in content:
		results.append({
			"id": int(element["id"]),
			"name": element["name"],
			#via https://github.com/alekstorm/pygravatar/blob/master/gravatar.py#L14
			"portrait": "http://gravatar.com/avatar/{}".format(md5(element["email"].strip().lower().encode()).hexdigest())
		})
	return results

def get_data2(): #offline debugging
	return [
		{"id": 0, "name": "A", "portrait": "/usr/share/ubuntu-html5-ui-toolkit/0.1/ambiance/img/avatar_contacts_list@8.png"},
		{"id": 1, "name": "B", "portrait": "/usr/share/ubuntu-html5-ui-toolkit/0.1/ambiance/img/avatar_contacts_list@8.png"},
		{"id": 2, "name": "C", "portrait": "/usr/share/ubuntu-html5-ui-toolkit/0.1/ambiance/img/avatar_contacts_list@8.png"}
	]


lib.js
function _http_json(method, url, params, callback)
{
	console.log("Requesting: " + url);
	var req = new XMLHttpRequest();
	req.open(method, url, true);
	req.onreadystatechange = function()
	{
		if(req.readyState != 4)
		{
			return;
		}
		if(req.status == 200) //TODO
		{
			console.log("Got reponse: " + req.responseText);
			console.log("Firing callback.");
			callback(JSON.parse(req.responseText));
		}
		else if(req.status == 204)
		{
			console.log("Got a successful response.");
			console.log("Firing callback.");
			callback();
		}
		else
		{
			console.log("Response: " + req.status);
			main.error(); //TODO: detailed message
		}
	};
	req.send(params ? new FormData(params) : null);
}

function get_users_data(url, callback)
{
	_http_json("GET", url + "/users.json", null, function(content)
	{
		var results = [];
		content.forEach(function(element)
		{
			results.push(_parseUser(element));
		});
		callback(results);
	});
}

function get_user_data(url, uid, callback)
{
	_http_json("GET", url + "/users/" + uid + ".json", null, function(content)
	{
		//assert(content["id"] == uid);
		var result = _parseUser(content);
		console.log(result);
		callback(result);
	});
}

function save_user_data(url, uid, name, balance, email, callback)
{
	var params = {
		"user[name]": name,
		"user[balance]": balance,
		"user[email]": email
	};
	_http_json("PATCH", url + "/users/" + uid + ".json", params, function(content)
	{
		callback(content); //TODO
	});
}

function add_user(url, name, balance, email, callback)
{
	var params = {
		"user[name]": name,
		"user[balance]": balance,
		"user[email]": email
	};
	_http_json("POST", url + "/users.json", params, function(content)
	{
		callback(_parseUser(content));
	});
}

function _parseUser(data)
{
	var result = {
		"id": data["id"],
		"name": data["name"],
		"balance": data["balance"],
		"email": data["email"],
		"portrait": get_gravatar_url(data["email"])
	};
	console.log(result);
	return result;
}

function get_gravatar_url(email)
{
	// see http://www.deluxeblogtips.com/2010/04/get-gravatar-using-only-javascript.html
	// MD5 (Message-Digest Algorithm) by WebToolkit
	var MD5=function(s){function L(k,d){return(k<<d)|(k>>>(32-d))}function K(G,k){var I,d,F,H,x;F=(G&2147483648);H=(k&2147483648);I=(G&1073741824);d=(k&1073741824);x=(G&1073741823)+(k&1073741823);if(I&d){return(x^2147483648^F^H)}if(I|d){if(x&1073741824){return(x^3221225472^F^H)}else{return(x^1073741824^F^H)}}else{return(x^F^H)}}function r(d,F,k){return(d&F)|((~d)&k)}function q(d,F,k){return(d&k)|(F&(~k))}function p(d,F,k){return(d^F^k)}function n(d,F,k){return(F^(d|(~k)))}function u(G,F,aa,Z,k,H,I){G=K(G,K(K(r(F,aa,Z),k),I));return K(L(G,H),F)}function f(G,F,aa,Z,k,H,I){G=K(G,K(K(q(F,aa,Z),k),I));return K(L(G,H),F)}function D(G,F,aa,Z,k,H,I){G=K(G,K(K(p(F,aa,Z),k),I));return K(L(G,H),F)}function t(G,F,aa,Z,k,H,I){G=K(G,K(K(n(F,aa,Z),k),I));return K(L(G,H),F)}function e(G){var Z;var F=G.length;var x=F+8;var k=(x-(x%64))/64;var I=(k+1)*16;var aa=Array(I-1);var d=0;var H=0;while(H<F){Z=(H-(H%4))/4;d=(H%4)*8;aa[Z]=(aa[Z]|(G.charCodeAt(H)<<d));H++}Z=(H-(H%4))/4;d=(H%4)*8;aa[Z]=aa[Z]|(128<<d);aa[I-2]=F<<3;aa[I-1]=F>>>29;return aa}function B(x){var k="",F="",G,d;for(d=0;d<=3;d++){G=(x>>>(d*8))&255;F="0"+G.toString(16);k=k+F.substr(F.length-2,2)}return k}function J(k){k=k.replace(/rn/g,"n");var d="";for(var F=0;F<k.length;F++){var x=k.charCodeAt(F);if(x<128){d+=String.fromCharCode(x)}else{if((x>127)&&(x<2048)){d+=String.fromCharCode((x>>6)|192);d+=String.fromCharCode((x&63)|128)}else{d+=String.fromCharCode((x>>12)|224);d+=String.fromCharCode(((x>>6)&63)|128);d+=String.fromCharCode((x&63)|128)}}}return d}var C=Array();var P,h,E,v,g,Y,X,W,V;var S=7,Q=12,N=17,M=22;var A=5,z=9,y=14,w=20;var o=4,m=11,l=16,j=23;var U=6,T=10,R=15,O=21;s=J(s);C=e(s);Y=1732584193;X=4023233417;W=2562383102;V=271733878;for(P=0;P<C.length;P+=16){h=Y;E=X;v=W;g=V;Y=u(Y,X,W,V,C[P+0],S,3614090360);V=u(V,Y,X,W,C[P+1],Q,3905402710);W=u(W,V,Y,X,C[P+2],N,606105819);X=u(X,W,V,Y,C[P+3],M,3250441966);Y=u(Y,X,W,V,C[P+4],S,4118548399);V=u(V,Y,X,W,C[P+5],Q,1200080426);W=u(W,V,Y,X,C[P+6],N,2821735955);X=u(X,W,V,Y,C[P+7],M,4249261313);Y=u(Y,X,W,V,C[P+8],S,1770035416);V=u(V,Y,X,W,C[P+9],Q,2336552879);W=u(W,V,Y,X,C[P+10],N,4294925233);X=u(X,W,V,Y,C[P+11],M,2304563134);Y=u(Y,X,W,V,C[P+12],S,1804603682);V=u(V,Y,X,W,C[P+13],Q,4254626195);W=u(W,V,Y,X,C[P+14],N,2792965006);X=u(X,W,V,Y,C[P+15],M,1236535329);Y=f(Y,X,W,V,C[P+1],A,4129170786);V=f(V,Y,X,W,C[P+6],z,3225465664);W=f(W,V,Y,X,C[P+11],y,643717713);X=f(X,W,V,Y,C[P+0],w,3921069994);Y=f(Y,X,W,V,C[P+5],A,3593408605);V=f(V,Y,X,W,C[P+10],z,38016083);W=f(W,V,Y,X,C[P+15],y,3634488961);X=f(X,W,V,Y,C[P+4],w,3889429448);Y=f(Y,X,W,V,C[P+9],A,568446438);V=f(V,Y,X,W,C[P+14],z,3275163606);W=f(W,V,Y,X,C[P+3],y,4107603335);X=f(X,W,V,Y,C[P+8],w,1163531501);Y=f(Y,X,W,V,C[P+13],A,2850285829);V=f(V,Y,X,W,C[P+2],z,4243563512);W=f(W,V,Y,X,C[P+7],y,1735328473);X=f(X,W,V,Y,C[P+12],w,2368359562);Y=D(Y,X,W,V,C[P+5],o,4294588738);V=D(V,Y,X,W,C[P+8],m,2272392833);W=D(W,V,Y,X,C[P+11],l,1839030562);X=D(X,W,V,Y,C[P+14],j,4259657740);Y=D(Y,X,W,V,C[P+1],o,2763975236);V=D(V,Y,X,W,C[P+4],m,1272893353);W=D(W,V,Y,X,C[P+7],l,4139469664);X=D(X,W,V,Y,C[P+10],j,3200236656);Y=D(Y,X,W,V,C[P+13],o,681279174);V=D(V,Y,X,W,C[P+0],m,3936430074);W=D(W,V,Y,X,C[P+3],l,3572445317);X=D(X,W,V,Y,C[P+6],j,76029189);Y=D(Y,X,W,V,C[P+9],o,3654602809);V=D(V,Y,X,W,C[P+12],m,3873151461);W=D(W,V,Y,X,C[P+15],l,530742520);X=D(X,W,V,Y,C[P+2],j,3299628645);Y=t(Y,X,W,V,C[P+0],U,4096336452);V=t(V,Y,X,W,C[P+7],T,1126891415);W=t(W,V,Y,X,C[P+14],R,2878612391);X=t(X,W,V,Y,C[P+5],O,4237533241);Y=t(Y,X,W,V,C[P+12],U,1700485571);V=t(V,Y,X,W,C[P+3],T,2399980690);W=t(W,V,Y,X,C[P+10],R,4293915773);X=t(X,W,V,Y,C[P+1],O,2240044497);Y=t(Y,X,W,V,C[P+8],U,1873313359);V=t(V,Y,X,W,C[P+15],T,4264355552);W=t(W,V,Y,X,C[P+6],R,2734768916);X=t(X,W,V,Y,C[P+13],O,1309151649);Y=t(Y,X,W,V,C[P+4],U,4149444226);V=t(V,Y,X,W,C[P+11],T,3174756917);W=t(W,V,Y,X,C[P+2],R,718787259);X=t(X,W,V,Y,C[P+9],O,3951481745);Y=K(Y,h);X=K(X,E);W=K(W,v);V=K(V,g)}var i=B(Y)+B(X)+B(W)+B(V);return i.toLowerCase()};
	var hash = MD5(email.trim().toLowerCase());
	return "http://gravatar.com/avatar/" + hash;
}

function get_drinks(url, callback)
{
	_http_json("GET", url + "/drinks.json", null, function(content)
	{
		var results = [];
		content.forEach(function(element)
		{
			results.push({
				"id": element["id"],
				"name": element["name"],
				"donation_recommendation": element["donation_recommendation"],
				"logo": element["logo_url"] ? (url + "/" + element["logo_url"]) : ""
			});
		});
		callback(results);
	});
}

function buy_drink(url, uid, did, callback)
{
	_http_json("GET", url + "/users/" + uid + "/buy.json?drink=" + did, null, function(content)
	{
		callback(content); //TODO
	});
}
